hs_core/tests/api/native/test_check_files.py

```diff
@@ -0,0 +1,395 @@
+
+        # TODO: how to eliminate this particular error.
+        # dumbpath = 'x' + shortpath
+        # dumbpath = self.res.short_id + "file1.txt"
+
+        # clean up after folder test
+        # ResourceFile.remove_folder(self.res, 'foo', self.user)
+
+        # delete resources to clean up
+        hydroshare.delete_resource(self.res.short_id)
+
+    def test_federated_root_path_logic(self):
+        """ a federated file path in the root folder has the proper state after state changes """
+        # resource should not have any files at this point
+        self.assertEqual(self.res.files.all().count(), 0,
+                         msg="resource file count didn't match")
+
+        # should succeed without errors
+        self.res.check_irods_files(stop_on_error=True)
+
+        # add one file to the resource
+        hydroshare.add_resource_files(self.res.short_id, self.test_file_1)
+
+        # should succeed without errors
+        self.res.check_irods_files(stop_on_error=True)
+
+        # resource should has only one file at this point
+        self.assertEqual(self.res.files.all().count(), 1,
+                         msg="resource file count didn't match")
+
+        # get the handle of the file created above
+        resfile = self.res.files.all()[0]
+
+        # cheat: set a fake federated path to test path logic
+        oldfedpath = self.res.resource_federation_path
+        oldpath = resfile.storage_path
+
+        # intentionally break path logic by setting an unused federation path
+        fedpath = "/myzone/home/myuser"
+        self.res.resource_federation_path = fedpath
+        self.res.save()
+        # must load changes into resfile from self.res before setting storage path
+        resfile.content_object.refresh_from_db()
+        resfile.set_storage_path('file1.txt', test_exists=False)
+
+        self.assertEqual(self.res.resource_federation_path, fedpath)
+        self.assertEqual(resfile.storage_path, get_path(resfile, 'file1.txt'))
+
+        # determine where that file should live; THIS IS FAKE
+        shortpath = os.path.join(fedpath, self.res.short_id, "data",
+                                 "contents", "file1.txt")
+
+        # intentionally break the resource file path
+        resfile.set_storage_path(shortpath, test_exists=False)
+        self.assertEqual(shortpath, resfile.storage_path)
+
+        self.assertEqual(resfile.file_folder, None)
+        self.assertEqual(resfile.storage_path, shortpath)
+
+        self.assertTrue(resfile.path_is_acceptable(shortpath, test_exists=False))
+
+        otherpath = os.path.join(fedpath, self.res.short_id, "data", "contents", "file2.txt")
+        resfile.path_is_acceptable(otherpath, test_exists=False)
+
+        # non-existent files should raise error
+        # This won't work because federation path is fake
+        # with self.assertRaises(ValidationError):
+        #     resfile.path_is_acceptable(otherpath, test_exists=True)
```

Comment:
If you want to test real (i.e., not fake) federation case, you can leverage iRODS local containers and add test case as done for federation unit tests, but I only see it as a nice to have, rather than a necessity at this point.